src/k8s/pkg/k8sd/controllers/csrsigning/validate_test.go

```diff
@@ -0,0 +1,228 @@
+				},
+			},
+			wantErr:    true,
+			errMessage: "failed to decrypt signature",
+		},
+		{
+			name: "Missing k8sd.io/node annotation",
+			csr: &certv1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"k8sd.io/signature": mustCreateEncryptedSignature(t, &key.PublicKey, csrPEM),
+					},
+				},
+				Spec: certv1.CertificateSigningRequestSpec{
+					Request:    []byte(csrPEM),
+					Username:   "system:node:valid-node",
+					Groups:     []string{"system:nodes"},
+					SignerName: "k8sd.io/kubelet-serving",
+					Usages:     []certv1.KeyUsage{certv1.UsageServerAuth, certv1.UsageDigitalSignature, certv1.UsageKeyEncipherment},
+				},
+			},
+			wantErr:    true,
+			errMessage: "k8sd.io/node annotation missing from CSR object",
+		},
+		{
+			name: "Invalid node name in CSR",
+			csr: &certv1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"k8sd.io/node":      "invalid-node!",
+						"k8sd.io/signature": mustCreateEncryptedSignature(t, &key.PublicKey, csrPEM),
+					},
+				},
+				Spec: certv1.CertificateSigningRequestSpec{
+					Request:    []byte(csrPEM),
+					Username:   "system:node:invalid-node!",
+					Groups:     []string{"system:nodes"},
+					SignerName: "k8sd.io/kubelet-serving",
+					Usages:     []certv1.KeyUsage{certv1.UsageServerAuth, certv1.UsageDigitalSignature, certv1.UsageKeyEncipherment},
+				},
+			},
+			wantErr:    true,
+			errMessage: "CSR has invalid node name",
+		},
+		{
+			name: "Invalid Signer Name",
+			csr: &certv1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"k8sd.io/signature": mustCreateEncryptedSignature(t, &key.PublicKey, csrPEM),
+						"k8sd.io/node":      "valid-node",
+					},
+				},
+				Spec: certv1.CertificateSigningRequestSpec{
+					Request:    []byte(csrPEM),
+					Username:   "system:node:valid-node",
+					Groups:     []string{"system:nodes"},
+					SignerName: "invalid-signer-name",
+					Usages:     []certv1.KeyUsage{certv1.UsageServerAuth, certv1.UsageDigitalSignature, certv1.UsageKeyEncipherment},
+				},
+			},
+			wantErr:    true,
+			errMessage: "CSR has unknown signerName",
+		},
+		{
+			name: "Invalid Usages",
+			csr: &certv1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"k8sd.io/signature": mustCreateEncryptedSignature(t, &key.PublicKey, csrPEM),
+						"k8sd.io/node":      "valid-node",
+					},
+				},
+				Spec: certv1.CertificateSigningRequestSpec{
+					Request:    []byte(csrPEM),
+					Username:   "system:node:valid-node",
+					Groups:     []string{"system:nodes"},
+					SignerName: "k8sd.io/kubelet-serving",
+					Usages:     []certv1.KeyUsage{certv1.UsageClientAuth}, // Invalid usages
+				},
+			},
+			wantErr:    true,
+			errMessage: "CSR usages",
```

Comment:
prefix "outputs" as "expectFoo", such that it is easy to tell them apart when revisiting the code in the future